covid/transform_utils.py

```diff
@@ -126,45 +206,81 @@ def generate_lags(df, column, num_lags=121, lag_timedelta=datetime.timedelta(day
     column_names.extend(lag_column_names)
 
     lags_df = pd.DataFrame(
-        index=pd.Index(data=states, name=STATE_FIELD), columns=column_names
+        index=pd.Index(data=locations, name=by_column), columns=column_names
     )
 
     today = pd.to_datetime(df[DATE_SOURCE_FIELD]).max()
     lags_df[DATE_SOURCE_FIELD] = today
-    for state in states:
+    for location in locations:
         # Start each state looking up today.
         date_to_lookup = today
 
         for lag in range(num_lags):
-            print(f"For field {column}, processing {state} for lag {lag}.")
+            print(f"For field {column}, processing {location} for lag {lag}.")
             # Lookup the historical entry.
             value = df.loc[
-                (df[STATE_FIELD] == state) & (df[DATE_SOURCE_FIELD] == date_to_lookup),
+                (df[by_column] == location) & (df[DATE_SOURCE_FIELD] == date_to_lookup),
                 column,
             ]
 
             if len(value) > 1:
                 raise ValueError("Too many or too few values returned.")
             elif len(value) == 1:
                 value = value.iloc[0]
-                lags_df.loc[state, column_name_formatter.format(lag)] = value
+                lags_df.loc[location, column_name_formatter.format(lag)] = value
 
             date_to_lookup = date_to_lookup - lag_timedelta
 
     lags_df = lags_df.reset_index()
     return lags_df
 
 
-def calculate_state_summary(transformed_df, columns):
+def compute_lagged_frame(
```

Comment:
As you said this logic is definitely complex but I'm sure it's way faster than the terrible for loops I had running! Should we potentially replace the other lagging parts of `transform` with this to also speed those up? (If we want to do that, we could potentially implement the text fixtures first so that we can tell if they worked successfully without changing the underlying data.)

Reply:
yeah I think the real improvement would be to apply transformations across states in parallel, which is maybe a larger refactor I didn't want to do here. I'm happy to add it in or just plug this function in on a state-by-state level